### ISSUE TITLE
Sharing resources (like a database pool) between tests

### ISSUE THREAD

**Issue by seagreen** [skogsbaer/HTF#49]
I have some tests that each need a database pool, but I'd rather not create an individual database pool within each test for performance reasons.

I'd like to write my tests like `prop_foo :: a -> Property`, and have some way to feed the result of my pool creation function `mkPool :: IO a` into each test.

Does HTF provide a way to do that and still use its TH test discovery? Even if one has to do discovery manually, is there a way to at least keep the same test settings and pretty results printing code as the rest of one's HTF tests?

**Comment by skogsbaer** [comment]
Good point! We are needing the feature as well. I have a plan how to implement it, but it will need some time.

**Comment by skogsbaer** [comment]
Here is a proposal for this feature.

From a users point of view, it would be nice to write something similar to this:

```haskell
-- define a TestResource for database connections
dbConResource :: TestResource
dbConResource = mkTestResource setup tearDown
    where
      setup = openDb "foo.db"
      tearDown = closeDb

-- write some test using this resource
test_userOk =
    withTestResource dbConResource $ \con ->
    ...

test_blub =
    withTestResource dbConResource $ \con ->
    ....
```

HTF then calls the `setup` function for the `TestResource` before the first test depending on the resource is run. HTF calls the `tearDown` function directly after the last test depending on resource has been run (prompt cleanup).

Internally, we need some restructuring in the TestManager (a test must somehow carry its resources to use) and in the ThreadPool (to allow for resource allocation and prompt deallocation).

**Comment by skogsbaer** [comment]
One more thing to consider: when implementing this feature, we should keep in mind #8 (benchmark support). For benchmarks, the setup function runs the reference benchmark (the teardown function does nothing). For this case, we also need to specify if tests should run concurrently to other tests or setup/teardown functions and if setup/teardown functions should run concurrently to other things.

**Comment by saurabhnanda** [comment]
Just re-checking if this has already been implemented, or this is still being considered. I looked at HTF for the first time today, and was confused about the lack of any examples where the tests are run in some sort of `ReaderT Env IO` monad. After reading this issue, my understanding is that each test is completely isolated form each other, and each test will have to manually execute `runReaderT  action env` within its own code. Did I understand this correctly?

**Comment by skogsbaer** [comment]
Yes, at the moment each test is run in isolation.
